src/main/webapp/app/complaints/complaint-response.service.ts

```diff
@@ -5,19 +5,65 @@ import { SERVER_API_URL } from 'app/app.constants';
 
 @Injectable({ providedIn: 'root' })
 export class ComplaintResponseService {
     private resourceUrl = SERVER_API_URL + 'api/complaint-responses';
 
-    constructor(private http: HttpClient) {}
+    constructor(private http: HttpClient, private accountService: AccountService) {}
 
-    create(complaintResponse: ComplaintResponse): Observable<EntityResponseType> {
+    /**
+     * Checks if a complaint response is locked for the currently logged in user
+     *
+     * A complaint response is never locked for the creator of the complaint response and for instructors
+     *
+     * @param complaintResponse complaint response to check the lock status for
+     * @param exercise exercise used to find out if currently logged in user is instructor
+     */
+    isComplaintResponseLockedForLoggedInUser(complaintResponse: ComplaintResponse, exercise: Exercise) {
+        if (complaintResponse.isCurrentlyLocked && complaintResponse.submittedTime === undefined) {
+            return !(complaintResponse.reviewer?.login === this.accountService.userIdentity?.login || this.accountService.isAtLeastInstructorForExercise(exercise));
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * Checks if the lock on a complaint response is active and if the currently logged in user is the creator of the lock
+     * @param complaintResponse complaint response to check
+     */
+    isComplaintResponseLockedByLoggedInUser(complaintResponse: ComplaintResponse) {
+        return !!(
+            complaintResponse.isCurrentlyLocked &&
+            complaintResponse.submittedTime === undefined &&
+            complaintResponse.reviewer?.login === this.accountService.userIdentity?.login
+        );
+    }
+
+    refreshLock(complaintId: number): Observable<EntityResponseType> {
+        return this.http
+            .post<ComplaintResponse>(`${this.resourceUrl}/complaint/${complaintId}/refresh-lock`, {}, { observe: 'response' })
+            .map((res: EntityResponseType) => this.convertDateFromServer(res));
+    }
+
+    removeLock(complaintId: number): Observable<HttpResponse<void>> {
+        return this.http.delete<void>(`${this.resourceUrl}/complaint/${complaintId}/remove-lock`, { observe: 'response' });
+    }
+
+    createLock(complaintId: number): Observable<EntityResponseType> {
+        return this.http
+            .post<ComplaintResponse>(`${this.resourceUrl}/complaint/${complaintId}/create-lock`, {}, { observe: 'response' })
+            .map((res: EntityResponseType) => this.convertDateFromServer(res));
+    }
+
+    resolveComplaint(complaintResponse: ComplaintResponse): Observable<EntityResponseType> {
         const copy = this.convertDateFromClient(complaintResponse);
         return this.http
-            .post<ComplaintResponse>(this.resourceUrl, copy, { observe: 'response' })
+            .put<ComplaintResponse>(`${this.resourceUrl}/complaint/${complaintResponse.complaint?.id}/resolve`, copy, { observe: 'response' })
```

Comment:
Imo we shouldn't send a request when `complaintRespone.complaint` is not set, i.e. when it is already clear at the client that the server will error (because the complaint ID will be null). On the other hand I don't know what our "best practices" here are. (Do we want to have the client send a bogus request so the server logs it, too?) I'd appreciate if you (or @krusche ) could clarify what to do.
Something like this:
```
if (!complaintResponse.complaint || complaintResponse.complaint!.id === 0) {
    // log error client side
    return;
}

const copy = this.convertDateFromClient(complaintResponse);
return this.http...
```

We could also just use `complaintResponse.complaint!.id` if we are sure the complaint exists.

Reply:
I think changing it to `complaintResponse.complaint!.id` is a good idea as it would fail in the client. I did it in 5cc66518682b6f84007d328b68e863716c4471ac